File: backend/schemas/invoices.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel, ConfigDict


class InvoiceItemCreate(BaseModel):
    description: str
    quantity: Decimal = Decimal("1")
    unit_price: Decimal
    revenue_account_id: Optional[uuid.UUID] = None


class InvoiceItemUpdate(BaseModel):
    description: Optional[str] = None
    quantity: Optional[Decimal] = None
    unit_price: Optional[Decimal] = None
    revenue_account_id: Optional[uuid.UUID] = None


class InvoiceItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    invoice_id: uuid.UUID
    revenue_account_id: Optional[uuid.UUID]
    description: str
    quantity: Decimal
    unit_price: Decimal
    amount: Decimal


class InvoiceCreate(BaseModel):
    client_id: uuid.UUID
    job_id: Optional[uuid.UUID] = None
    status: str = "draft"
    discount: Decimal = Decimal("0")
    tax: Decimal = Decimal("0")
    due_date: Optional[date] = None


class InvoiceUpdate(BaseModel):
    status: Optional[str] = None
    discount: Optional[Decimal] = None
    tax: Optional[Decimal] = None
    due_date: Optional[date] = None


class PaymentCreate(BaseModel):
    amount: Decimal
    payment_date: date
    account_id: uuid.UUID
    notes: Optional[str] = None


class PaymentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    invoice_id: uuid.UUID
    amount: Decimal
    payment_date: date
    account_id: uuid.UUID
    notes: Optional[str]
    created_at: datetime


class InvoiceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_id: Optional[uuid.UUID]
    client_id: uuid.UUID
    status: str
    subtotal: Decimal
    discount: Decimal
    tax: Decimal
    total: Decimal
    balance_due: Decimal
    requires_review: bool
    due_date: Optional[date]
    sent_at: Optional[datetime]
    paid_at: Optional[datetime]
    created_at: datetime
    items: list[InvoiceItemOut] = []
    payments: list[PaymentOut] = []

File: frontend/src/schemas/invoices.ts
import { z } from 'zod'

export const PaymentSchema = z.object({
  id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  amount: z.string(),
  payment_date: z.string(),
  account_id: z.string().uuid(),
  notes: z.string().nullable(),
  created_at: z.string(),
})
export type Payment = z.infer<typeof PaymentSchema>
export const PaymentListSchema = z.array(PaymentSchema)

const InvoiceItemSchema = z.object({
  id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  revenue_account_id: z.string().uuid().nullable(),
  description: z.string(),
  quantity: z.string(),
  unit_price: z.string(),
  amount: z.string(),
})

export const InvoiceSchema = z.object({
  id: z.string().uuid(),
  job_id: z.string().uuid().nullable(),
  client_id: z.string().uuid(),
  status: z.enum(['draft', 'sent', 'partially_paid', 'paid']),
  subtotal: z.string(),
  discount: z.string(),
  tax: z.string(),
  total: z.string(),
  balance_due: z.string(),
  requires_review: z.boolean(),
  due_date: z.string().nullable(),
  sent_at: z.string().nullable(),
  paid_at: z.string().nullable(),
  created_at: z.string(),
  items: z.array(InvoiceItemSchema),
  payments: z.array(PaymentSchema),
})
export type Invoice = z.infer<typeof InvoiceSchema>
export const InvoiceListSchema = z.array(InvoiceSchema)
